### PR TITLE
utilities: improve margin specification type and add space-spec type

### PR DESCRIPTION
**Pull request by dkochmanski**
Revert to the old behavior and add a comment which explains why it is
not checking for REALP.

### CODE REVIEW

**Comment by dkochmanski** [comment]
```
20:02 < scymtym> jackdaniel: new test failure for NORMALIZE-MARGIN-SPEC.VALIDATION: https://jenkins-experimental.bob.ci.cit-ec.net/blue/organizations/jenkins/cl-mcclim-multibranch-with-doc-ci-docker-mcclim-nightly/detail/fix-tf-margin-parse-space/1/tests
20:26 < jackdaniel> scymtym: spec in the test is indeed invalid, but I think we should remove this class of tests. otherwise we may want to define a type which will satisfy a new predicate space-parsable-p (which type would make sense), but I won't get to it before next week
20:27 < scymtym> jackdaniel: i would prefer the second solution
20:27 < scymtym> or something amounting to even more work :)
```

**Comment by dkochmanski** [comment]
I've added space-spec type with a few improvements for parse-space.

File: Core/clim-basic/extended-streams/text-formatting.lisp [hunk 1]
```diff
@@ -18,33 +18,6 @@
...
-        for value = (getf plist edge)
-        do
-           (typecase value
-             (null (setf (getf plist edge) (getf defaults edge)))
-             (atom (setf (getf plist edge) `(:relative ,value)))
-             (list #| do nothing |#))
-        finally
-           (check-type plist margin-spec)
-           (return plist)))
-
```
Comment (scymtym) [inline Core/clim-basic/extended-streams/text-formatting.lisp]: I think this was fine in `text-formatting.lisp`. If you want to move the functions, please also move the test.

File: Core/clim-basic/utils.lisp [hunk 2]
```diff
@@ -538,20 +538,47 @@ STREAM in the direction DIRECTION."
                     (:vertical   (values 0 1))))
               (/ value (sqrt (+ (* dx dx) (* dy dy))))))))))))
 
+(defun valid-margin-spec-p (margins)
```
Comment (scymtym) [inline Core/clim-basic/utils.lisp]: I would squash the following commits into one:
* `b7ff860b5 margin-spec: move utility to utils and validate the value precisely`
* `dec980b88 parse-space: add corresponding type and enchance the implementation` (typo in "enchance")
* `4b66aacd4 text-formatting: margin-spec-p: do not assume value to be a number`

File: Core/clim-basic/utils.lisp [hunk 3]
```diff
@@ -468,10 +468,27 @@ in KEYWORDS removed."
...
+;;; add a special handling for the non-stream panes. -- jd 2019-11-02
+
+(deftype space-spec ()
+  `(or real
+       string
+       character
+       function
+       (cons real
+             (cons (member :character :line :point :pixel :mm)
+                   null))))
```
Comment (scymtym) [inline Core/clim-basic/utils.lisp]: Generators for testing `parse-space' directly or for improving the `normalize-margin-spec` test:

```lisp
(defun gen-unit ()
  (gen-one-element :character :line :point :pixel :mm))

(defun gen-invalid-space-spec ()
  (let ((value (gen-integer :min 0 :max 100))
        (unit (gen-unit)))
    (lambda ()
      (case (random 5)
        (0 :invalid-space)
        (1 #C(1 2))
        (2 (list (funcall value) :invalid-unit))
        (3 (list :invalid-real (funcall unit)))
        (4 (list (funcall value) (funcall unit) 3))))))

(defun gen-invalid-margin-spec-element ()
  (let ((anchor (gen-one-element :relative :absolute))
        (space (gen-invalid-space-spec)))
    (lambda ()
      (case (random 3)
        (0 :foo)
        (1 (list (funcall anchor) 1 2))
        (2 (list (funcall anchor) (funcall space)))))))
```

File: Core/clim-basic/utils.lisp (hunk 2 above)
Comment (dkochmanski) [reply Core/clim-basic/utils.lisp]: I'd rather keep parse-space type in a separate commit, the other two are squashed (with moved tests).

File: Core/clim-basic/utils.lisp (hunk 3 above)
Comment (dkochmanski) [reply Core/clim-basic/utils.lisp]: added to utils.lisp test

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 1 above)
Comment (dkochmanski) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: moved the test. I've kept the file but it doesn't have tests anymore.

**Comment by dkochmanski** [comment]
all remarks are addressed (see comments)

File: Tests/text-formatting.lisp [hunk 4]
```diff
@@ -3,63 +3,4 @@
...
-     (case (random 3)
-       (0 :foo)
-       (1 (list (funcall anchor) :foo))
-       (2 (list (funcall anchor) 1 2))))))
-
-(defun gen-invalid-margin-spec ()
-  (let ((edge (gen-one-element :left :right :top :bottom))
-        (element (gen-invalid-margin-spec-element)))
-    (lambda ()
-      (list (funcall edge) (funcall element)))))
```
Comment (scymtym) [inline Tests/text-formatting.lisp]: This seems to have been deleted when moving the code. The test fails without it.

Comment (dkochmanski) [reply Tests/text-formatting.lisp]: fixed

Comment (dkochmanski) [reply Tests/text-formatting.lisp]: fixed

**Pull request merged**